File: JS/src/utils/warehouseExtensions.js
class WarehouseExtensions {
    static getWarehouseVolume(warehouse) {
        if (warehouse.maxCapacity) {
            return warehouse.maxCapacity;
        }
        const { height, width, length } = warehouse;
        if (height && width && length) {
            return height * width * length;
        }
        throw new Error("Warehouse does not have enough information to calculate volume.");
    }

    static getItemVolume(item) {
        return item.volume || 0; 
    }

    static getVolumeOccupiedOnDay(warehouse, day) {
        const normalizedDay = new Date(day).setHours(0, 0, 0, 0);
        return warehouse.items
            .filter(item => {
                const normalizedStartDate = new Date(item.startDate).setHours(0, 0, 0, 0);
                const normalizedEndDate = new Date(item.endDate).setHours(0, 0, 0, 0);
                return (
                    item.isActive !== false &&
                    normalizedDay >= normalizedStartDate &&
                    normalizedDay <= normalizedEndDate
                );
            })
            .reduce((total, item) => total + WarehouseExtensions.getItemVolume(item), 0);
    }
}

module.exports = WarehouseExtensions;

File: WarehouseProjectDotNET/WarehouseProject/Services/WarehouseExtensions.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using WarehouseProject.Models;

namespace WarehouseProject.Services
{
    public static class WarehouseExtensions
    {
        public static double GetWarehouseVolume(this Warehouse warehouse)
        {
            return warehouse.MaxCapacity.Height * warehouse.MaxCapacity.Width * warehouse.MaxCapacity.Length;
        }

        public static double GetItemVolume(this Item item)
        {
            return item.ItemHeight * item.ItemWidth * item.ItemLength;
        }

        /// <summary>
        /// Returns the sum of the volumes of all items that are active and overlap the given day.
        /// </summary>
        public static double GetVolumeOccupiedOnDay(this Warehouse warehouse, DateTime day)
        {
            return warehouse.Items
                .Where(i => i.IsActive && day >= i.StartDate && day <= i.EndDate)
                .Sum(i => i.GetItemVolume());
        }
    }
}
